chaining of metadata

sugarforever opened the pull request:
Notice that, the schema files are to be submitted shortly.

sssion committed:
add ut

sssion committed:
fix bug of speedtest

sssion committed:
update ci

sssion committed:
add ut

sssion committed:
fix bug of waitCheckList

sssion committed:
fix bug of register

sssion committed:
update to latest go-legs and datastore, delete subscribe

sssion committed:
add schema

sssion committed:
add schema and update core linksystem, legs hook

bsjohnson01 committed:
update sdk

bsjohnson01 committed:
fix metadata block persistence issue

bsjohnson01 committed:
fix metadata block persistence issue

masih commented on pkg/types/schema/schema.ipldsch:
```diff
@@ -0,0 +1,8 @@
+type Metadata struct {
+	# Previous metadata.
+	PreviousID optional Link_Metadata
+	Payload    Bytes
```
I think we also need providers to sign the metadata. Pando would then need to verify the signature to assert that the metadata is indeed produced by the provider

sssion committed:
update unittest, mockpando and clean code

sssion committed:
update meta schema, add signature, provider and modify the previousID as nullable. Verify meta when arrived in linksystem.

sssion committed:
fix bug of default estuary gateway

sssion committed:
update mock_provider and add unittest about fetch linked meta data and backup linked meta data

sssion committed:
update metadata_manage for generating car files.
1. only back up for registered providers
2. use backup config
3.  three parameters for backup
(1)backup generate interval
(2)back up estuary interval
(3)estuary deal check interval

bsjohnson01 committed:
add field provider into appendMetadata method

sugarforever commented on pkg/api/v1/http/pando.go:
```diff
@@ -25,27 +24,27 @@ func (a *API) registerPando() {
+//	}
+//
+//	ctx.JSON(http.StatusOK, types.NewOKResponse("subscribe success", nil))
+//}
```
这部分注释代码可以删除

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
-			return err
-		}
-		return &buffer, committer, nil
+			// If it is an advertisement.
```
删除注释

sugarforever commented on pkg/legs/core_test.go:
```diff
@@ -2,59 +2,134 @@ package legs_test
+			_, err = pando.BS.Get(ctx, cids[i])
+			So(err, ShouldBeNil)
+		}
+		//for {
```
删除注释代码

sugarforever commented on pkg/legs/core.go:
```diff
@@ -2,230 +2,211 @@ package legs
+	// SyncPrefix used to track the latest sync in datastore.
+	SyncPrefix = "/sync/"
 	// PubSubTopic used for legs transportation
 	PubSubTopic = "PandoPubSub"
```
@sssion 确认一下后续的代码改动是否会改topic为/pando/v0.0.1?

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
+}
+
+// providerFromAd reads the provider ID from an advertisement
+func providerFromAd(ad schema.Metadata) (peer.ID, error) {
```
函数名，变量及注释中的advertisement，建议改成metadata。

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
-			return
-		}
+func verifyMetadata(n ipld.Node) (schema.Metadata, peer.ID, error) {
+	meta, err := decodeAd(n)
```
decodeAd函数名，变量及注释中的advertisement，建议改成metadata相关的名称。

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
+func verifyMetadata(n ipld.Node) (schema.Metadata, peer.ID, error) {
+	meta, err := decodeAd(n)
+	if err != nil {
+		log.Errorw("Cannot decode advertisement", "err", err)
```
advertisement改成metadata

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
+		return nil, peer.ID(""), err
+	}
+
+	// Get provider ID from advertisement.
```
advertisement改成metadata

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
+	// Get provider ID from advertisement.
+	provID, err := providerFromAd(meta)
+	if err != nil {
+		log.Errorw("Cannot get provider from advertisement", "err", err)
```
advertisement改成metadata

sugarforever commented on pkg/legs/link_system.go:
```diff
@@ -26,70 +28,125 @@ func MkLinkSystem(bs blockstore.Blockstore) ipld.LinkSystem {
+func providerFromAd(ad schema.Metadata) (peer.ID, error) {
+	provider, err := ad.FieldProvider().AsString()
+	if err != nil {
+		return peer.ID(""), fmt.Errorf("cannot read provider from advertisement: %w", err)
```
advertisement改成metadata

sugarforever commented on pkg/legs/ratelimiter.go:
```diff
@@ -1,79 +1,78 @@
-	}
-	log.Debugf("request %d canceled", request)
-}
+//func (l *Core) cancelRequest(request graphsync.RequestID) {
```
删除注释代码

sugarforever commented on pkg/lotus/discovery.go:
```diff
@@ -46,74 +37,74 @@ func NewDiscoverer(gateway string) (*Discoverer, error) {
-		Balance:  balance.Int,
-	}, nil
-}
+//func (d *Discoverer) getMinerPeerAddr(minerInfo miner.MinerInfo) (peer.AddrInfo, error) {
```
删除注释代码

sugarforever commented on pkg/lotus/discovery_test.go:
```diff
@@ -26,18 +26,20 @@ func TestDiscoverMock(t *testing.T) {
 	})
 }
+
+//func TestDiscover(t *testing.T) {
```
删除注释代码

sugarforever commented on pkg/metadata/backup.go:
```diff
@@ -17,37 +17,32 @@ import (
+type BackupSystem struct {
+	backupCfg *option.Backup
+	apiKey    string
+	//fileName       string
```
fileName不需要了，可以删除

sugarforever commented on pkg/metadata/metadata_manage.go:
```diff
@@ -27,13 +28,15 @@ import (
-	BackFileName           = "backup-%d.car"
-	BackupCheckNumInterval = time.Second * 60
+	SnapShotDuration = time.Second * 5
+	//BackupMaxInterval      = time.Second * 10
```
请删除注释代码

sugarforever commented on pkg/metadata/metadata_manage.go:
```diff
@@ -27,13 +28,15 @@ import (
+	BackupTmpPath     string
+	BackFileName      = "backup-%s-%d.car"
+	BackupGenInterval = time.Minute
+	//BackupCheckNumInterval = time.Second * 60
```
请删除注释代码

sugarforever commented on pkg/metadata/metadata_manage.go:
```diff
@@ -65,27 +68,29 @@ type MetaManager struct {
-	cid      cid.Cid
-	isBackup bool
-}
+//type backupRecord struct {
```
请删除注释代码

sugarforever commented on pkg/metadata/metadata_manage.go:
```diff
@@ -169,105 +173,63 @@ func (mm *MetaManager) GetUpdateOut() <-chan map[peer.ID]*types.ProviderState {
+					log.Errorf("failed to export backup car for provider:%s\nerr:%s",
+						info.AddrInfo.ID.String(), err.Error())
+				}
+				err = mm.regstry.RegisterOrUpdate(ctx, info.AddrInfo.ID, lastSyncCid)
```
mm.exportMetaCar返回err != nil的情况下，是否还需要继续mm.regstry.RegisterOrUpdate操作？

sugarforever commented on pkg/registry/registry.go:
```diff
@@ -350,18 +337,69 @@ func (r *Registry) loadPersistedProviders() (int, error) {
-				log.Warnf("failed to subscribe: %s, err: %s", peerID, err.Error())
-			}
-		}
+		//if r.core == nil {
```
请删除注释代码

sugarforever commented on pkg/registry/registry.go:
```diff
@@ -350,18 +337,69 @@ func (r *Registry) loadPersistedProviders() (int, error) {
+}
+
+// RegisterOrUpdate attempts to register an unregistered provider, or updates
+// the addresses and latest advertisement of an already registered provider.
```
advertisement改为metadata

sugarforever commented on pkg/statetree/state_tree.go:
```diff
@@ -284,9 +289,9 @@ func (st *StateTree) GetSnapShot(sscid cid.Cid) (shot *types2.SnapShot, err erro
-	if height < 0 {
-		return nil, fmt.Errorf("height must be positive")
-	}
+	//if height < 0 {
```
提前判断height < 0，可以直接返回，免掉调用st.GetSnapShotCidList()的开销。是否考虑保留这个if。

sssion replied:
传参是uint64

sugarforever commented on sdk/pkg/provider/dag.go:
```diff
@@ -88,32 +95,41 @@ func (p *DAGProvider) Close() error {
-func (p *DAGProvider) PushMany(nodes []ipldFormat.Node) error {
-	ctx, cancel := context.WithTimeout(context.Background(), p.PushTimeout)
-	defer cancel()
+	logger.Infow("Publishing advertisement in pubsub channel", "cid", c)
```
advertisement改成metadata

sugarforever commented on sdk/pkg/provider/dag.go:
```diff
@@ -88,32 +95,41 @@ func (p *DAGProvider) Close() error {
+	c, err := p.PushLocal(ctx, metadata)
 	if err != nil {
-		return err
+		return cid.Undef, fmt.Errorf("failed to publish advertisement locally: %s", err)
```
advertisement改成metadata

sugarforever commented on test/mock/discovery.go:
```diff
@@ -17,21 +17,21 @@ type mockDiscoverer struct {
-	minerAddr      = "/ip4/127.0.0.1/tcp/9999"
-	minerAddr2     = "/ip4/127.0.0.2/tcp/9999"
+	exceptID = "12D3KooWK7CTS7cyWi51PeNE3cTjS2F2kDCZaQVU4A5xBmb9J1do"
+	//	trustedID  = "12D3KooWSG3JuvEjRkSxt93ADTjQxqe4ExbBwSkQ9Zyk1WfBaZJF"
```
删除这部分的注释代码
